File: app/src/androidTest/java/com/chrisprime/primestationonecontrol/utilities/SpoonScreenshotUtilities.java
package com.chrisprime.primestationonecontrol.utilities;

import android.app.Activity;

import com.squareup.spoon.Spoon;

import timber.log.Timber;

public class SpoonScreenshotUtilities {
    public static Activity sScreenshotActivity;

    public static void screenshot(String tag) {
        screenshot(tag, sScreenshotActivity);
    }

    public static void screenshot(String tag, String className, String methodName) {
        screenshot(tag, sScreenshotActivity, className, methodName);
    }

    public static void screenshot(String tag, Activity currentActivity) {
        screenshot(tag, currentActivity, null, null);
    }

    public static void screenshot(String tag, Activity currentActivity, String className, String methodName) {
        Timber.d(".screenshot(%s) requested!", tag);
        if (currentActivity != null) {
            if (StringUtilities.INSTANCE.isEmpty(className) || StringUtilities.INSTANCE.isEmpty(methodName)) {
                Spoon.screenshot(currentActivity, tag);
            } else {
                Spoon.screenshot(currentActivity, tag, className, methodName);
            }
            saveDatabase(tag, currentActivity);
        } else {
            Timber.w(".screenshot(%s): Unable to proceed, currentActivity is null!", tag);
        }
    }

    public static void saveDatabase(String tag, Activity currentActivity) {
        Timber.d(".saveDatabase(%s) requested!", tag);
//        File databaseFile = new ContextWrapper(currentActivity).getDatabasePath(DATABASE_NAME);
//        try {
//            Spoon.save(currentActivity, databaseFile); //Grab the database upon failure so it can easily be downloaded and examined from the Spoon test reoprt too!
//            Timber.d(".screenshot(%s): successfully saved database snapshot!", tag);
//        } catch (RuntimeException e) {
//            Timber.w(".screenshot(%s): Unable to save database from path %s - does not yet exist?", tag, databaseFile.getAbsolutePath());
//        }
    }
}

File: app/src/androidTest/java/com/chrisprime/primestationonecontrol/utilities/SpoonScreenshotUtilities.kt
package com.chrisprime.primestationonecontrol.utilities

import android.app.Activity
import android.text.TextUtils
import com.jraska.falcon.FalconSpoon
import com.squareup.spoon.Spoon
import timber.log.Timber
import java.io.File
import java.net.URLEncoder


@Suppress("unused")
object SpoonScreenshotUtilities {

    @JvmStatic fun screenshot(tag: String) {
        screenshot(tag, TestUtilities.currentActivity)
    }

    @JvmStatic fun screenshot(tag: String, className: String, methodName: String) {
        screenshot(tag, TestUtilities.currentActivity, className, methodName)
    }

    @JvmStatic fun screenshot(tag: String, currentActivity: Activity?) {
        screenshot(tag, currentActivity, null, null)
    }

    @JvmStatic private fun screenshot(tag: String, currentActivity: Activity?, className: String?, methodName: String?) {
        val sanitizedTag = safeStringToScreenshotName(tag)
        Timber.d(".screenshot(%s) requested!  Sanitized tag to %s", tag, sanitizedTag)
        try {
            if (currentActivity != null) {
                if (TextUtils.isEmpty(className) || TextUtils.isEmpty(methodName)) {
                    FalconSpoon.screenshot(currentActivity, sanitizedTag)
                } else {
                    FalconSpoon.screenshot(currentActivity, sanitizedTag, className, methodName)
                }
                saveData(sanitizedTag, currentActivity)
            } else {
                Timber.w(".screenshot(%s): Unable to proceed, currentActivity is null!", tag)
            }
        } catch (e: RuntimeException) {
            Timber.e(e, ".screenshot(%s): Unable to proceed, we might not have permission to write to the external storage: %s, is this API 23+?", tag, e.message)
        }
    }

    @JvmStatic private fun saveData(tag: String, currentActivity: Activity) {
//        Timber.d(".saveData(%s) requested!", tag)
//        val contextWrapper = ContextWrapper(currentActivity)
//        val databaseFile = contextWrapper.getDatabasePath(SchemaValidator.DATABASE_NAME)
//        val realmFile = File(String.format("%s%s%s", contextWrapper.filesDir, File.separator, "db.realm"))
        val sharedPrefsFile = File(String.format("%s%s%s%s%s%s%s%s%s%s", File.separator, "data", File.separator, "data", File.separator, "com.chrisprime.primestationonecontrol.debug", File.separator, "shared_prefs", File.separator, "com.chrisprime.primestationonecontrol.debug_preferences.xml"))

//        saveFileToReport(currentActivity, databaseFile, tag)
//        saveFileToReport(currentActivity, realmFile, tag)
        saveFileToReport(currentActivity, sharedPrefsFile, tag)
    }

    @JvmStatic private fun saveFileToReport(currentActivity: Activity, file: File, tag: String) {
        try {
            Spoon.save(currentActivity, file) //Grab the database so it can easily be downloaded and examined from the Spoon test reoprt too!
            Timber.d(".screenshot(%s): successfully saved %s snapshot!", tag, file.absolutePath)

        } catch (e: RuntimeException) {
            Timber.w(".screenshot(%s): Unable to save from path %s - does not yet exist? %s", tag, file.absolutePath, e.message)
        }
    }

    @JvmStatic private fun safeStringToScreenshotName(string: String) = URLEncoder.encode(string, "UTF-8").replace('%', '_').replace('+', '_').replace('.', '_')
}
